File: src/multiple_turtlebots_sim/scripts/image_extractor.py
#! /usr/bin/python
import rospy
# ROS Image message
from sensor_msgs.msg import Image
# ROS Image message -> OpenCV2 image converter
from cv_bridge import CvBridge, CvBridgeError
# OpenCV2 for saving an image

import message_filters

import cv2
import time

rgb_path = "/home/ayumi/Documents/tohoku_uni/CLOVERs/images/1-image/"
depth_path = "/home/ayumi/Documents/tohoku_uni/CLOVERs/images/depth-image/"

rospy.init_node('image_listener', anonymous=True)

robot_number = rospy.get_param("~robot_number")
total_num = rospy.get_param("~total")
img_number=robot_number

# Instantiate CvBridge
bridge = CvBridge()

def callback(rgb,depth):
    global img_number
    
    cv2_img = bridge.imgmsg_to_cv2(rgb, "bgr8")
    cv2_img = cv2.resize(cv2_img, dsize=(512, 384))
    cv2.imwrite(rgb_path + str(img_number) + ".jpg", cv2_img)
    
    cv2_img_depth=bridge.imgmsg_to_cv2(depth,"passthrough")
    cv2.imwrite(depth_path + str(img_number) +".png", cv2_img_depth)
    # Depth image はpng!
    img_number += total_num

    time.sleep(1/5);

# Define your image topic

rgb_topic = "/robot"+str(robot_number)+"/camera/rgb/image_raw"
depth_topic = "/robot"+str(robot_number)+"/camera/depth/image_raw"
# # To store the images from camera
rgb_sub=message_filters.Subscriber(rgb_topic,Image)
depth_sub=message_filters.Subscriber(depth_topic,Image)

# ts=message_filters.ApproximateTimeSynchronizer([rgb_sub,depth_sub],10,0.1,allow_headerless=False)
ts=message_filters.ApproximateTimeSynchronizer([rgb_sub,depth_sub],10,0.1,allow_headerless=False)
ts.registerCallback(callback)
# rospy.Subscriber(image_topic, Image, rgb_callback)
rospy.spin()

File: src/cpp/src/loop_closure/image_extractor.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encoding.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/CameraInfo.h>
#include <std_msgs/Int32.h>

using namespace sensor_msgs;
using namespace message_filters;

static const std::string OPENCV_WINDOW = "Image window";

class ImageConverter{
    private:
    int robot_number;

    int rgb_index{0};
    int depth_index{0};

    std::string rgb_path{"/home/ayumi/Documents/tohoku_uni/images/1_rgb"};
    std::string depth_path{"/home/ayumi/Documents/tohoku_uni/images/1_depth"};

    std::string all_rgb{"/home/ayumi/Documents/tohoku_uni/images/all_rgb"};
    std::string all_rgb{"/home/ayumi/Documents/tohoku_uni/images/all_depth"};

    ros::NodeHandle n;
    image_transport::ImageTransport it_;
    image_transport::Subscriber rgb_sub;
    image_transport::Subscriber depth_sub;
    // Check if there is 4 img or not
    image_transport::Publisher commander;

    public:
    ImageConverter(int num):it_(n),robot_number(num){

        rgb_sub=it_.subscribe("robot";std::to_string(robot_number)+"/camera/color/image_raw",1,
        &ImageConverter::rgbCB,this);
        depth_sub=it_.subscribe("robot";std::to_string(robot_number)+"/camera/depth/image_raw",1,
        &ImageConverter::depthCB,this);

        commander.advertise("robot";std::to_string(robot_number)+"command_RO_estimate",1);

    }

    void rgbCB(const sensor_msgs::ImageConstPtr& msg){
        rgb_index+=1;
        cv_bridge::CvImagePtr cv_ptr;
        cv_ptr=cv_bridge::toCvCopy(msg,sensor_msgs::image_encodings::BGR8);
        cv::imwrite(std::to_string(rgb_index)+".jpg",cv_ptr->image);

    }

    }

void callback(const ImageConstPtr &image, const CameraInfoConstPtr &cam_info)
{
    // Solve all of perception here...
}

int main(int argc, char **argv)
{
    ros::init(argc,argv,"image_extractor_node");
    ros::NodeHandle pnh("~");
    int robot_No;
    pnh.getParam("robot_No",robot_No);

    ImageConverter(robot_No);
    
    message_filters::Subscriber<Image> image_sub(nh, "image", 1);
    message_filters::Subscriber<CameraInfo> info_sub(nh, "camera_info", 1);
    TimeSynchronizer<Image, CameraInfo> sync(image_sub, info_sub, 10);
    sync.registerCallback(boost::bind(&callback, _1, _2));

    ros::spin();

    return 0;
}


int main(int argc,char **argv){
    ros::init(argc,argv,"image_extractor_node");
    ros::NodeHandle pnh("~");
    int robot_No;
    pnh.getParam("robot_No",robot_No);

    ImageConverter(robot_No);

    ros::spin();

    return 0;
}